### DISCUSSION TITLE
GitHub Action to build pyodide packages

### DISCUSSION THREAD

**Discussion by jobovy** [pyodide/pyodide#2719] [Show and tell]
I created a simple GitHub Action to build `pyodide` wheels of Python packages in a workflow:

https://github.com/jobovy/pyodide-buildpackage-action

The intended use cases are:
1. To build `pyodide` wheels of the latest development version for packages in the `pyodide` distribution or of a new release of such a package to have it available before it gets updated in `pyodide`.
2. To build `pyodide` wheels for packages that aren't in the distribution

Because of use case 1., the basic flow of the action is that it takes a `meta.yaml` file (e.g., that for a package already in `pyodide`) and replaces the URL with another URL (default: the commit that triggered the workflow that uses this action). The action is implemented in a modular fashion that allows on to (relatively) easily build multiple packages (e.g., all of the dependencies of a package).

Some comments about the implementation:
- Implementation is simply: check out a `pyodide` version/tag/branch, copy in an updated `meta.yaml`, build the package and store the resulting wheel.
- I use the tagged, pre-built Docker image, but because GitHub actions removes everything at the Docker image's `WORKDIR` and replaces it with its own working directory, the action can't use the built version of `pyodide`, but has to re-build it from scratch. If there was some way to avoid `WORKDIR` in the Docker image, then this could potentially be avoided.
- Because building `pyodide` takes a long time, I implemented caching so subsequent build are sped up. This works well, but the cache is very large, a few GB. So it fills up much of the cache available to a repository.
- Because dependencies can take a long time to build and aren't really necessary if they already exist within `pyodide` (mainly `scipy`), by default I only build the package and any `library/sharedlibrary` dependencies. That does mean that if a package needs another non-`library/sharedlibrary` dependency during build time that this would fail. I'm not sure whether that's actually ever the case, but I imagine it is. In that case, specifying which are build-time dependencies and which are run-time dependencies in the `meta.yaml` file would be useful.

I hope this is useful to other users and I would appreciate any feedback!

**Comment by hoodmane** [comment]
Thanks @jobovy! I think this can be simplified a lot soon. In fact the set of files needed to build packages is pretty small, under 2mb. Some partial progress on this:
https://github.com/pyodide/pyodide/pull/2550

I have been working on adding CI to pyo3, and there I manage to build Pyodide packages out of tree, not in the docker image, and without building Pyodide.
https://github.com/PyO3/setuptools-rust/pull/244

Do you have an example repo where you are using this buildpackage-action? I could make a PR on your repo trying to prototype better out of tree builds.

**Reply by jobovy** [reply]
Thanks for the feedback!

> Do you have an example repo where you are using this buildpackage-action? I could make a PR on your repo trying to prototype better out of tree builds.

I'm using it in [this branch](https://github.com/jobovy/galpy/tree/add-pyodide-wheels) to build wheels of my `galpy` code so I can use the latest version for an interactive example in the documentation (like `numpy` has). The relevant workflow is [here](https://github.com/jobovy/galpy/blob/add-pyodide-wheels/.github/workflows/build_pyodide_wheels.yml). Note that I have to use a more complicated version of the action I wrote, because I'm building against `pyodide` 0.20.0 and thus have to add the GSL by hand as well (which wasn't in 0.20.0; I have to use 0.20.0, because I'm using `jupyterlite` in my docs and that doesn't support the latest `pyodide`) as well as the `galpy` patches. It would probably be better for you to work with an example that doesn't need as much hacking to work, but I don't have such an example currently.

**Reply by hoodmane** [reply]
Yeah that sounds too complicated.

**Reply by jobovy** [reply]
Actually, if you just build an updated version of `galpy` against the latest `pyodide` then it's as simple as replacing the URL in `galpy`'s `meta.yaml`, so it's very easy. I do this in [this repository](https://github.com/jobovy/test-pyodide-buildpackage-action) that I used to test the action, so you could use that one if you want? The `meta.yaml` that's in that repository is just the exact one from `pyodide`, so the only 'action' is in the [workflow file](https://github.com/jobovy/test-pyodide-buildpackage-action/blob/main/.github/workflows/build-pyodide.yml).

**Comment by coatless** [comment]
This looks like the action is stuck in time with respect to Pyodide 20, would it be worthwhile to bring it back up to code or just to spend time developing a separate action for building pyodide packages?

**Reply by agriyakhetarpal** [reply]
I have had an idea for setting up an official action at https://github.com/pyodide/pyodide-actions/issues/12 which would work with any version of Pyodide, nightlies fromhttps://github.com/pyodide/pyodide-build-environment-nightly/releases, and with its own constraints and necessary, but I haven't taken the time to attempt it. PRs to do it would be welcome, and I'd be happy to help review and merge it!

Otherwise, you should be able to use `cibuildwheel` with `CIBW_PLATFORM: pyodide` and `--platform pyodide` to build and test packages, which should provide a convenient interface for now.

**Reply by jobovy** [reply]
I don't think there's much point to updating this action, I haven't even been using it myself even. And it seems like just using `cibuildwheel` is a much easier way to accomplish the same thing. An official action would be great though, especially if it could then give dependabot updates to keep versions up-to-date (which is hard to do manually and then suddenly one finds oneself many versions behind...)
